src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java

```diff
@@ -121,40 +148,107 @@ public void auditFinished(AuditEvent event) {
 
     @Override
     public void fileStarted(AuditEvent event) {
-        writer.println("<file name=\"" + encode(event.getFileName()) + "\">");
+        fileMessages.put(event.getFileName(), new FileMessages());
     }
 
     @Override
     public void fileFinished(AuditEvent event) {
+        final String fileName = event.getFileName();
+        final FileMessages errorEvents = fileMessages.get(fileName);
+
+        synchronized (writerLock) {
+            if (errorEvents != null) {
+                writeFileOpeningTag(fileName);
+
+                for (AuditEvent errorEvent : errorEvents.errors) {
+                    writeFileError(errorEvent);
+                }
+                for (Throwable exception : errorEvents.exceptions) {
+                    writeException(exception);
+                }
+            }
+
+            writeFileClosingTag();
```

Comment:
@rnveach This is similar to this (https://github.com/checkstyle/checkstyle/pull/4933#discussion_r133356702). Previously, `fileStarted` prints opening tag, and `fileFinished` prints closing tag. In case, if a developer forgets to open or close tag, the `XMLLogger` will not print corresponding.

Probably, there was a reason for it, that is why I decided not to change existing tests. However, in my opition, `XMLLogger` should open (or close) missed tags